src/circuit/gadgets/constraints.rs

```diff
@@ -427,6 +464,148 @@ pub fn must_be_simple_bit(x: &Boolean) -> AllocatedBit {
     }
 }
 
+// Enforce div and mod operation for U64. We need to show that
+// arg1 = q * arg2 + r, such that 0 <= r < arg2.
+pub fn enforce_u64_div_mod<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    g: &GlobalAllocations<F>,
+    cond: Boolean,
+    arg1: &AllocatedPtr<F>,
+    arg2: &AllocatedPtr<F>,
+    store: &Store<F>,
+) -> Result<(AllocatedPtr<F>, AllocatedPtr<F>), SynthesisError> {
+    let arg1_u64 = match arg1.hash().get_value() {
+        Some(v) => v.to_u64_unchecked(),
+        None => 0, // Blank and Dummy
+    };
+    let arg2_u64 = match arg2.hash().get_value() {
+        Some(v) => v.to_u64_unchecked(),
+        None => 0, // Blank and Dummy
+    };
+
+    let (q, r) = if arg2_u64 != 0 {
+        (arg1_u64 / arg2_u64, arg1_u64 % arg2_u64)
+    } else {
+        (0, 0)
+    };
+
+    let r_ptr = store.get_u64(r);
+    let alloc_r =
+        AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "u64 remainder"), store, || Ok(&r_ptr))?;
+
+    let q_ptr = store.get_u64(q);
+    let alloc_q =
+        AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "u64 quotient"), store, || Ok(&q_ptr))?;
+
+    let arg1_u64_ptr = store.get_u64(arg1_u64);
+    let alloc_arg1 = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "alloc arg1"), store, || {
+        Ok(&arg1_u64_ptr)
+    })?;
+
+    let arg2_u64_ptr = store.get_u64(arg2_u64);
+    let alloc_arg2 = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "alloc arg2"), store, || {
+        Ok(&arg2_u64_ptr)
+    })?;
+
+    // a = b * q + r
+    let product_u64mod = mul(
+        &mut cs.namespace(|| "product(q,arg2)"),
+        alloc_q.hash(),
+        alloc_arg2.hash(),
+    )?;
+    let sum_u64mod = add(
+        &mut cs.namespace(|| "sum remainder mod u64"),
+        &product_u64mod,
+        alloc_r.hash(),
+    )?;
+    let u64mod_decomp = alloc_equal(
+        &mut cs.namespace(|| "check u64 mod decomposition"),
+        &sum_u64mod,
+        alloc_arg1.hash(),
+    )?;
+    let b_is_zero = alloc_is_zero(&mut cs.namespace(|| "b is zero"), arg2.hash())?;
+    let b_is_not_zero_and_cond = Boolean::and(
+        &mut cs.namespace(|| "b is not zero and cond"),
+        &b_is_zero.not(),
+        &cond,
+    )?;
+    enforce_implication(
+        &mut cs.namespace(|| "enforce u64 mod decomposition"),
+        &b_is_not_zero_and_cond,
+        &u64mod_decomp,
+    )?;
+
+    enforce_less_than_bound(
+        &mut cs.namespace(|| "remainder in range b"),
+        cond,
+        alloc_r.clone(),
+        alloc_arg2.clone(),
+    )?;
+
+    let arg1_u64_tag = alloc_equal(
+        &mut cs.namespace(|| "arg1 u64 tag"),
+        alloc_arg1.tag(),
+        &g.u64_tag,
+    )?;
+    let arg2_u64_tag = alloc_equal(
+        &mut cs.namespace(|| "arg2 u64 tag"),
+        alloc_arg2.tag(),
+        &g.u64_tag,
+    )?;
+    enforce_true(&mut cs.namespace(|| "check arg1 u64 tag"), &arg1_u64_tag)?;
+    enforce_true(&mut cs.namespace(|| "check arg2 u64 tag"), &arg2_u64_tag)?;
```

Comment:
As above, and just to be very explicit about the comments I'm making here, this check seems superfluous because these allocated tag values aren't even returned.

The general comment I would make applies throughout this PR and suggests a slight disconnection with regard to how we should approach these questions. I think we've discussed this before but might not yet be totally aligned. Specifically, I think we should avoid dealing with Lurk-specific constructs (like `AllocatedPtr`) in most general circuit implementation. We shouldn't be using Lurk data types to 'compute' or 'reason about' the underlying logic except in the cases where we're actually explicitly dealing with how types interact with Lurk evaluation. The previous example related to booleans, where I suggested we should only 'wrap' `Boolean` values by turning them into Lurk booleans (`T` or `NIL`) when required in evaluation. We shouldn't be doing 'boolean logic' using Lurk `AllocatedPtr`s. The same applies here to field-element logic, I think.

I note that this could change in the future if we have a more robust set of gadgets to *efficiently* take advantage of explicit type information — but I don't think our current model does that, so using 'Lurk' values when unnecessary muddies the water.

It's my belief that `constraints.rs` shouldn't contain any references to `AllocatedPtr` (or `Ptr`, etc.).

Reply:
done